tasks/components.py

```diff
@@ -0,0 +1,186 @@
+                comment_block = comment_block[:i]
+            except ValueError:
+                pass
+            return ''.join(comment_block).strip() + '\n'
+        else:
+            comment_block = []
+
+
+def has_type_component(content):
+    return any(l.startswith('type Component interface') for l in content)
+
+
+def get_components_and_bundles(ctx):
+    ok = True
+    components = []
+    bundles = []
+    res = ctx.run('git ls-files comp/', hide=True)
+    for file in res.stdout.splitlines():
+        if file.endswith("/component.go"):
+            content = list(open(file, "r"))
+            if not has_type_component(content):
+                print(f"** {file} does not define a Component interface; skipping")
+                ok = False
+                pass
+
+            path = file[: -len('/component.go')]
+            team = find_team(content)
+            doc = find_doc(content)
+
+            if team is None:
+                print(f"** {file} does not name a responsible team")
+                ok = False
+
+            components.append(Component(path, doc, team))
+
+        elif file.endswith("/bundle.go"):
+            content = list(open(file, "r"))
+            if has_type_component(content):
+                print(f"** {file} defines a Component interface (bundles should not do so)")
+                ok = False
+                pass
+
+            path = file[: -len('/bundle.go')]
+            team = find_team(content)
+            doc = find_doc(content)
+
+            if team is None:
+                print(f"** {file} does not name a responsible team")
+                ok = False
+
+            bundles.append(Bundle(path, doc, team, []))
+
+    # assign components to bundles
+    bundles = [Bundle(b.path, b.doc, b.team, [c for c in components if c.path.startswith(b.path)]) for b in bundles]
+
+    # look for un-bundled components
+    for c in components:
+        if not any(c in b.components for b in bundles):
+            print(f"** component {c.path} is not in any bundle")
+            ok = False
+
+    return sorted(bundles), ok
+
+
+def make_components_md(bundles):
+    pkg_root = 'github.com/DataDog/dd-agent-comp-experiments/'
+    yield '# Agent Components'
+    yield ''
+    yield 'This file lists all components defined in this repository, with their package summary.'
+    yield 'Click the links for more documentation.'
+    yield ''
+    for b in bundles:
+        yield f'## [{b.path}](https://pkg.go.dev/{pkg_root}{b.path}) (Component Bundle)'
+        yield ''
+        yield f'*Datadog Team*: {b.team}'
+        yield ''
+        yield b.doc
+        for c in b.components:
+            yield f'### [{c.path}](https://pkg.go.dev/{pkg_root}{c.path})'
+            yield ''
+            if c.team != b.team:
+                yield f'*Datadog Team*: {c.team}'
+                yield ''
+            yield c.doc
+
+
+def make_codeowners(codeowners_lines, bundles):
+    codeowners_lines = codeowners_lines.__iter__()
```

Comment:
Why do you need to do this?

Reply:
I call `next(..)` explicitly, so I need an iterator and not a list.

Reply:
Sorry, that was in an earlier version -- but, the `for` loops are meant to iterate sequentially, not to each begin at the beginning of the list.